### DISCUSSION TITLE
Making updates a little safer on RHEL?

### DISCUSSION THREAD

**Discussion by io7m** [openzfs/zfs#16799] [General]
Hello!

I run a single RHEL 9 server (Rocky, to be precise) for a small business. We used to run Fedora, but found it too fast moving, and almost every `dnf update` would end up breaking `zfs` due to version incompatibility. Moving to Rocky/RHEL obviously helped matters a lot as it's much slower moving, and we'd largely been running `dnf update` fairly often expecting that we'd only be receiving security updates. This seemed to be fine, nothing ever changed to the point that `zfs` would stop being available after a reboot.

Then, of course, this morning we upgraded, rebooted, and found `zfs` [no longer working](https://github.com/openzfs/zfs/issues/16746). A workaround in that ticket (`dnf update --enablerepo=zfs-testing zfs`) got us running again.

Is there some way we can ensure that we don't perform an update if the resulting system won't have working `zfs`? I was less vigilantly watching `dnf update` as I wasn't expecting a `zfs` breakage, but if there was any kind of warning/error, it apparently passed by unnoticed. We really only run `dnf update` for security patches.

**Comment by mskarbek** [comment]
https://github.com/zeet-cc/zfs-rpms/blob/main/kernel-zfs.spec - I'm using this as a safeguard. It's an empty package that locks in `kmod-zfs` and `kernel-core` together, which prevents undesirable kernel updates. Works perfectly fine on both Fedora and RHEL.

**Reply by io7m** [reply]
Interesting! Could you elaborate a bit on this?

I'm not completely familiar with the `rpm` format, but I can see how the constraints in that package spec might have the desired effect. Is this something you have to maintain yourself, at every zfs update?

**Comment by deajan** [comment]
I usually use this oneliner before rebooting after a zfs upgrade to check whether all kernels have zfs module:
```
for kernel in $(find /boot -name "vmlinuz*x86*" | awk -F"-" '{gsub("/boot/vmlinuz-",""); print $0}' | sort -n); do if [ "$(find /usr/lib/modules/$kernel -iname "zfs.ko*")" != "" ]; then HAS_ZFS=ZFS_OK; else HAS_ZFS=NO_ZFS; fi; if [ $(uname -r) == $kernel ]; then echo "CURRENT $HAS_ZFS $kernel"; else echo "OTHER   $HAS_ZFS $kernel"; fi; done
```
